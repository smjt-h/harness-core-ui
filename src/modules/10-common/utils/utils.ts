/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export const isCommunityPlan = (): boolean => false

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined
