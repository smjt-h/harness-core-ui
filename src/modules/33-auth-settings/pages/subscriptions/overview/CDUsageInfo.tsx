/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { Layout, PageError } from '@wings-software/uicore'
import { useStrings } from 'framework/strings'
import { useGetUsageAndLimit } from '@common/hooks/useGetUsageAndLimit'
import { ContainerSpinner } from '@common/components/ContainerSpinner/ContainerSpinner'
import { ModuleName } from 'framework/types/ModuleName'
import UsageInfoCard, { ErrorContainer } from './UsageInfoCard'

const ActiveInstanceCard: React.FC<{ subscribedIns: number; activeIns: number; displayName?: string }> = ({
  subscribedIns,
  activeIns,
  displayName
}) => {
  const { getString } = useStrings()

  const leftHeader = getString('common.subscriptions.usage.srvcInst')
  const tooltip = getString('common.subscriptions.usage.cdSITooltip')
  const rightHeader = displayName || getString('common.subscriptions.usage.last30days')
  const hasBar = true
  const leftFooter = getString('common.subscribed')
  const rightFooter = getString('common.subscribed')
  const props = {
    subscribed: subscribedIns,
    usage: activeIns,
    leftHeader,
    tooltip,
    rightHeader,
    hasBar,
    leftFooter,
    rightFooter
  }
  return <UsageInfoCard {...props} />
}

const ActiveServices: React.FC<{ subscribedService: number; activeService: number; displayName?: string }> = ({
  subscribedService,
  activeService,
  displayName
}) => {
  const { getString } = useStrings()

  const leftHeader = getString('common.subscriptions.usage.services')
  const tooltip = getString('common.subscriptions.usage.cdServiceTooltip')
  const rightHeader = displayName || getString('common.subscriptions.usage.last30days')
  const hasBar = true
  const leftFooter = getString('total')
  const props = {
    subscribed: subscribedService,
    usage: activeService,
    leftHeader,
    tooltip,
    rightHeader,
    hasBar,
    leftFooter
  }
  return <UsageInfoCard {...props} />
}

const CDUsageInfo: React.FC = () => {
  const { limitData, usageData } = useGetUsageAndLimit(ModuleName.CD)
  const isLoading = limitData.loadingLimit || usageData.loadingUsage

  if (isLoading) {
    return <ContainerSpinner />
  }

  const { usageErrorMsg, refetchUsage, usage } = usageData
  const { limitErrorMsg, refetchLimit, limit } = limitData

  if (usageErrorMsg) {
    return (
      <ErrorContainer>
        <PageError message={usageErrorMsg} onClick={() => refetchUsage?.()} />
      </ErrorContainer>
    )
  }

  if (limitErrorMsg) {
    return (
      <ErrorContainer>
        <PageError message={limitErrorMsg} onClick={() => refetchLimit?.()} />
      </ErrorContainer>
    )
  }
  return (
    <Layout.Horizontal spacing="large">
      <ActiveInstanceCard
        subscribedIns={limit?.cd?.totalServiceInstances || 0}
        activeIns={usage?.cd?.activeServiceInstances?.count || 0}
        displayName={usage?.cd?.activeServiceInstances?.displayName}
      />
      <ActiveServices
        subscribedService={limit?.cd?.totalWorkload || 0}
        activeService={usage?.cd?.activeServices?.count || 0}
        displayName={usage?.cd?.activeServices?.displayName}
      />
    </Layout.Horizontal>
  )
}

export default CDUsageInfo
