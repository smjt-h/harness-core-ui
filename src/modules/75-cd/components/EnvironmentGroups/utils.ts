/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo } from 'lodash-es'
import type {
  EnvironmentGroupFilterProperties,
  EnvironmentGroupResponseDTO,
  GetEnvironmentGroupListQueryParams
} from 'services/cd-ng'

export function cleanData(values: EnvironmentGroupResponseDTO): EnvironmentGroupResponseDTO {
  const {
    name: prevName,
    identifier: prevIdentifier,
    description: prevDescription,
    tags,
    orgIdentifier,
    projectIdentifier,
    envIdentifiers
  } = values

  const name = prevName?.toString().trim()
  const identifier = prevIdentifier?.toString().trim()
  const description = prevDescription?.toString().trim()
  return {
    name,
    identifier,
    description,
    tags: defaultTo(tags, {}),
    orgIdentifier,
    projectIdentifier,
    envIdentifiers
  } as any
}

export enum SortFields {
  LastUpdatedAt = 'lastModifiedAt',
  AZ09 = 'AZ09',
  ZA90 = 'ZA90',
  Name = 'name'
}

export enum Sort {
  DESC = 'DESC',
  ASC = 'ASC'
}

export enum EnvironmentGroupDetailsTab {
  CONFIGURATION = 'CONFIGURATION',
  ENVIRONMENTS = 'ENVIRONMENTS'
}

export interface EnvironmentGroupListQueryParams
  extends Partial<
    Omit<GetEnvironmentGroupListQueryParams, 'sort' | 'getDefaultFromOtherRepo' | 'envGroupIdentifiers'>
  > {
  filters?: EnvironmentGroupFilterProperties
}
