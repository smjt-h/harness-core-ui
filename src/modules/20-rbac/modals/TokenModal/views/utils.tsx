/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getReadableDateTime } from '@common/utils/dateUtils'

export const DATE_FORMAT = 'MM/DD/YYYY'

export const hasNoExpiryDate = (value?: number): boolean => {
  return getReadableDateTime(value, DATE_FORMAT) === '12/31/2099'
}

export const getSelectedExpiration = (value?: number): string => {
  if (!value) {
    return '30'
  } else if (hasNoExpiryDate(value)) {
    return '-1'
  }
  return 'custom'
}

export const getSelectedExpirationDate = (value: string): string => {
  switch (value) {
    case '30':
    case '90':
    case '180':
      return getReadableDateTime(new Date().setDate(new Date().getDate() + parseInt(value)), DATE_FORMAT)
    case '-1':
      return '12/31/2099'
    default:
      return ''
  }
}

export const getExpirationDate = (value?: number): string => {
  if (value) {
    return getReadableDateTime(value, DATE_FORMAT)
  } else {
    return getSelectedExpirationDate(getSelectedExpiration(value))
  }
}
