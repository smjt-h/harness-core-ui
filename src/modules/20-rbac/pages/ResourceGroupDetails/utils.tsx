/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import produce from 'immer'
import type { SelectOption } from '@wings-software/uicore'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import { isDynamicResourceSelector, SelectionType } from '@rbac/utils/utils'
import type {
  ResourceSelector,
  ResponseResourceTypeDTO,
  ScopeSelector,
  ResourceGroupV2,
  ResourceSelectorV2,
  ResourceGroupV2Request,
  ResourceFilter
} from 'services/resourcegroups'
import { Scope } from '@common/interfaces/SecretsInterface'
import type { UseStringsReturn } from 'framework/strings'
import RbacFactory from '@rbac/factories/RbacFactory'
import { getScopeFromDTO } from '@common/components/EntityReference/EntityReference'

export enum SelectorScope {
  CURRENT = 'CURRENT',
  INCLUDE_CHILD_SCOPES = 'INCLUDE_CHILD_SCOPES',
  CUSTOM = 'CUSTOM'
}

enum ValidatorTypes {
  BY_RESOURCE_IDENTIFIER = 'BY_RESOURCE_IDENTIFIER',
  BY_RESOURCE_TYPE = 'BY_RESOURCE_TYPE',
  BY_RESOURCE_TYPE_INCLUDING_CHILD_SCOPES = 'BY_RESOURCE_TYPE_INCLUDING_CHILD_SCOPES'
}

export const getSelectedResourcesMap = (
  resourceTypes: ResourceType[],
  resourceFilter?: ResourceFilter
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()
  if (resourceFilter?.includeAllResources) {
    resourceTypes.forEach(resourceType => {
      map.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    })
  } else {
    resourceFilter?.resources?.map(resourceSelector => {
      if (resourceSelector.identifiers?.length) {
        map.set(resourceSelector.resourceType as ResourceType, resourceSelector.identifiers)
      } else {
        map.set(resourceSelector.resourceType as ResourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      }
    })
  }

  return map
}

export const validateResourceSelectors = (value: ResourceSelector[]): ResourceType[] => {
  return value
    .filter(val => {
      if (!isDynamicResourceSelector(val.type)) {
        if (val.identifiers?.length < 1) {
          return true
        }
      }
      return false
    })
    .map(val => val.resourceType)
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelectorV2[] => {
  const resourceSelectors: ResourceSelectorV2[] = []
  map.forEach((value, key) => {
    if (isDynamicResourceSelector(value)) {
      resourceSelectors.push({
        resourceType: key
      })
    } else {
      resourceSelectors.push({
        resourceType: key,
        identifiers: value as string[]
      })
    }
  })
  return resourceSelectors
}

export const computeResourceMapOnChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  resourceType: ResourceType,
  isAdd: boolean,
  identifiers?: string[]
): void => {
  if (identifiers) {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, identifiers)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          if (Array.isArray(draft.get(resourceType))) {
            draft.set(
              resourceType,
              (draft.get(resourceType) as string[]).filter(el => !identifiers.includes(el))
            )
          }
          if (draft.get(resourceType)?.length === 0) {
            draft.delete(resourceType)
          }
        })
      )
    }
  } else {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.delete(resourceType)
        })
      )
    }
  }
}

export const computeResourceMapOnMultiChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  types: ResourceType[],
  isAdd: boolean
): void => {
  if (isAdd) {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      })
    )
  } else {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.delete(resourceType)
        })
      })
    )
  }
}

export const getScopeLabelFromApi = (
  getString: UseStringsReturn['getString'],
  scope: Scope,
  resourceGroup: ResourceGroupV2
): string => {
  const selectorScope = getScopeType(resourceGroup)
  const dropDownItems = getScopeDropDownItems(scope, getString)
  const option = dropDownItems.filter(item => item.value === selectorScope)
  return option.length ? option[0].label : ''
}

export const getScopeDropDownItems = (scope: Scope, getString: UseStringsReturn['getString']): SelectOption[] => {
  switch (scope) {
    case Scope.PROJECT:
      return [{ label: getString('rbac.scopeItems.projectOnly'), value: SelectorScope.CURRENT }]
    case Scope.ORG:
      return [
        { label: getString('rbac.scopeItems.orgOnly'), value: SelectorScope.CURRENT },
        { label: getString('rbac.scopeItems.orgAll'), value: SelectorScope.INCLUDE_CHILD_SCOPES }
      ]
    case Scope.ACCOUNT:
    default:
      return [
        { label: getString('rbac.scopeItems.accountOnly'), value: SelectorScope.CURRENT },
        { label: getString('rbac.scopeItems.accountAll'), value: SelectorScope.INCLUDE_CHILD_SCOPES }
      ]
  }
}

export const getSelectionType = (resourceGroup?: ResourceGroupV2): SelectionType => {
  if (resourceGroup?.resourceFilter?.includeAllResources) {
    return SelectionType.ALL
  }
  return SelectionType.SPECIFIED
}
export const getScopeType = (resourceGroup?: ResourceGroupV2): SelectorScope => {
  if (resourceGroup?.includedScopes?.length) {
    for (const scope of resourceGroup?.includedScopes) {
      if (getScopeFromDTO(resourceGroup) === getScopeFromDTO(scope)) {
        return scope.filter === 'INCLUDING_CHILD_SCOPES' ? SelectorScope.INCLUDE_CHILD_SCOPES : SelectorScope.CURRENT
      } else {
        return SelectorScope.CUSTOM
      }
    }
  }

  return SelectorScope.CURRENT
}

export const getFormattedDataForApi = (
  data: ResourceGroupV2,
  selectionType: SelectionType,
  includedScopes: ScopeSelector[],
  resourceSelectors: ResourceSelectorV2[]
): ResourceGroupV2Request => {
  return {
    resourceGroup: {
      ...data,
      resourceFilter:
        selectionType === SelectionType.ALL
          ? {
              includeAllResources: true
            }
          : {
              resources: resourceSelectors,
              includeAllResources: false
            },
      includedScopes: includedScopes === [] ? getIncludedScopes(SelectorScope.CURRENT, data) : includedScopes
    }
  }
}

export const getFilteredResourceTypes = (
  data: ResponseResourceTypeDTO | null,
  selectedScope: SelectorScope
): ResourceType[] => {
  const resourceTypes = data?.data?.resourceTypes || []
  return resourceTypes?.reduce((acc: ResourceType[], value) => {
    if (
      value.name &&
      RbacFactory.isRegisteredResourceType(value.name as ResourceType) &&
      ((selectedScope === SelectorScope.CURRENT && value.validatorTypes?.includes(ValidatorTypes.BY_RESOURCE_TYPE)) ||
        selectedScope !== SelectorScope.CURRENT)
    ) {
      acc.push(value.name as ResourceType)
    }
    return acc
  }, [])
}

export const cleanUpResourcesMap = (
  resourceTypes: ResourceType[],
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  selectionType: SelectionType
): Map<ResourceType, string | string[]> => {
  const types = [...resourceTypes]
  selectedResourcesMap.forEach((_value, key) => {
    if (!types.includes(key)) {
      selectedResourcesMap.delete(key)
    } else {
      types.splice(types.indexOf(key), 1)
      selectedResourcesMap.set(key, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    }
  })
  if (selectionType === SelectionType.ALL) {
    types.map(resourceType => {
      selectedResourcesMap.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    })
  }
  return selectedResourcesMap
}

export const getDefaultIncludedScope = (
  accountIdentifier: string,
  orgIdentifier?: string,
  projectIdentifier?: string,
  includedScopes?: ScopeSelector[]
): ScopeSelector[] => {
  if (includedScopes?.length) {
    return includedScopes
  }
  return [
    {
      accountIdentifier,
      orgIdentifier,
      projectIdentifier,
      filter: 'EXCLUDING_CHILD_SCOPES'
    }
  ]
}

export const getIncludedScopes = (type: SelectorScope, resourceGroup: ResourceGroupV2): ScopeSelector[] => {
  const { accountIdentifier, orgIdentifier, projectIdentifier } = resourceGroup
  switch (type) {
    case SelectorScope.CUSTOM:
      //Replace with custom scope once it is added
      return []
    case SelectorScope.INCLUDE_CHILD_SCOPES:
      return [
        {
          accountIdentifier,
          orgIdentifier,
          projectIdentifier,
          filter: 'INCLUDING_CHILD_SCOPES'
        }
      ]
    default:
      return [
        {
          accountIdentifier,
          orgIdentifier,
          projectIdentifier,
          filter: 'EXCLUDING_CHILD_SCOPES'
        }
      ]
  }
}
