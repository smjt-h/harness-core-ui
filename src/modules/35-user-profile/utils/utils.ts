/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@wings-software/uicore'
import { ValueType } from '@secrets/components/TextReference/TextReference'
import type { SCMData } from '@user-profile/modals/SourceCodeManager/views/SourceCodeManagerForm'
import type { SourceCodeManagerAuthentication, SourceCodeManagerDTO } from 'services/cd-ng'
import { setSecretField } from '@secrets/utils/SecretField'

export enum ConnectionType {
  HTTPS = 'HTTPS',
  HTTP = 'Http',
  SSH = 'Ssh'
}

export enum AuthTypes {
  USERNAME_PASSWORD = 'UsernamePassword',
  USERNAME_TOKEN = 'UsernameToken',
  KERBEROS = 'Kerberos',
  SSH_KEY = 'SSH_KEY',
  AWSCredentials = 'AWSCredentials'
}

export interface SourceCodeType {
  text: string
  value: SourceCodeTypes
  icon: IconName
}

export enum SourceCodeTypes {
  BITBUCKET = 'BITBUCKET',
  GITHUB = 'GITHUB',
  GITLAB = 'GITLAB',
  AZURE_REPO = 'AZURE_REPO',
  AWS_CODE_COMMIT = 'AWS_CODE_COMMIT'
}

export const selectedValueToTypeMap: Record<SourceCodeTypes, SourceCodeTypes> = {
  [SourceCodeTypes.BITBUCKET]: SourceCodeTypes.BITBUCKET,
  [SourceCodeTypes.GITHUB]: SourceCodeTypes.GITHUB,
  [SourceCodeTypes.GITLAB]: SourceCodeTypes.GITLAB,
  [SourceCodeTypes.AWS_CODE_COMMIT]: SourceCodeTypes.AWS_CODE_COMMIT,
  [SourceCodeTypes.AZURE_REPO]: SourceCodeTypes.AZURE_REPO
}

export const getIconBySCM = (item: SourceCodeTypes): IconName => {
  switch (item) {
    case SourceCodeTypes.BITBUCKET:
      return 'bitbucket-blue'
    case SourceCodeTypes.GITHUB:
      return 'github'
    case SourceCodeTypes.GITLAB:
      return 'service-gotlab'
    case SourceCodeTypes.AWS_CODE_COMMIT:
      return 'service-aws-code-deploy'
    case SourceCodeTypes.AZURE_REPO:
      return 'service-azure'
    default:
      return 'bitbucket'
  }
}

export const getAuthentication = (values: SCMData): SourceCodeManagerAuthentication | undefined => {
  switch (values.authType) {
    case AuthTypes.USERNAME_PASSWORD:
      return {
        type: ConnectionType.HTTP,
        spec: {
          type: AuthTypes.USERNAME_PASSWORD,
          spec: {
            ...(values.username?.type === ValueType.TEXT
              ? { username: values.username.value }
              : { usernameRef: values.username?.value }),
            passwordRef: values.password?.referenceString
          }
        }
      }
    case AuthTypes.USERNAME_TOKEN:
      return {
        type: ConnectionType.HTTP,
        spec: {
          type: AuthTypes.USERNAME_TOKEN,
          spec: {
            ...(values.username?.type === ValueType.TEXT
              ? { username: values.username.value }
              : { usernameRef: values.username?.value }),
            tokenRef: values.accessToken?.referenceString
          }
        }
      }
    default:
      return undefined
  }
}

const getBitBucketFormData = async (
  sourceCodeManagerData: SourceCodeManagerDTO,
  accountIdentifier: string
): Promise<SCMData> => {
  const { name, authentication } = sourceCodeManagerData
  return {
    name: name,
    authType: authentication?.spec?.type,
    password: (await setSecretField(authentication?.spec?.spec?.passwordRef, {
      accountIdentifier
    })) as SCMData['password'],
    username: {
      value: authentication?.spec?.spec?.username || authentication?.spec?.spec?.usernameRef,
      type: authentication?.spec?.spec?.usernameRef ? ValueType.ENCRYPTED : ValueType.TEXT
    }
  }
}

const getGithubFormData = async (
  sourceCodeManagerData: SourceCodeManagerDTO,
  accountIdentifier: string
): Promise<SCMData> => {
  const { name, authentication } = sourceCodeManagerData
  return {
    name: name,
    authType: authentication?.spec?.type,
    accessToken: (await setSecretField(
      authentication?.spec?.spec?.tokenRef || authentication?.spec?.apiAccess?.spec?.tokenRef,
      { accountIdentifier }
    )) as SCMData['accessToken'],
    username: {
      value: authentication?.spec?.spec?.username || authentication?.spec?.spec?.usernameRef,
      type: authentication?.spec?.spec?.usernameRef ? ValueType.ENCRYPTED : ValueType.TEXT
    }
  }
}

export const getFormDataBasedOnSCMType = (
  sourceCodeManagerData: SourceCodeManagerDTO,
  accountIdentifier: string
): Promise<SCMData | undefined> => {
  switch (sourceCodeManagerData.type) {
    case SourceCodeTypes.BITBUCKET:
      return getBitBucketFormData(sourceCodeManagerData, accountIdentifier)
    case SourceCodeTypes.GITHUB:
      return getGithubFormData(sourceCodeManagerData, accountIdentifier)
    default:
      return Promise.resolve(undefined)
  }
}

export const getDefaultSCMType = (
  supportedSCMS: SourceCodeType[],
  currentSCM?: SourceCodeManagerDTO['type']
): SourceCodeType => {
  return supportedSCMS.find(item => item.value === currentSCM) || supportedSCMS[0]
}

export const getDefaultSelected = (type?: SourceCodeTypes): AuthTypes | undefined => {
  switch (type) {
    case SourceCodeTypes.GITHUB:
      return AuthTypes.USERNAME_TOKEN
    case SourceCodeTypes.BITBUCKET:
    case SourceCodeTypes.GITLAB:
    case SourceCodeTypes.AZURE_REPO:
      return AuthTypes.USERNAME_PASSWORD
    case SourceCodeTypes.AWS_CODE_COMMIT:
      return AuthTypes.AWSCredentials
    default:
      return undefined
  }
}
