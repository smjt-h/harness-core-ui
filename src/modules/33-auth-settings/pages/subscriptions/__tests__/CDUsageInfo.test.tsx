/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'

import CDUsageInfo from '../overview/CDUsageInfo'

jest.mock('@common/hooks/useGetUsageAndLimit', () => {
  return {
    useGetUsageAndLimit: () => {
      return useGetUsageAndLimitReturnMock
    }
  }
})
const useGetUsageAndLimitReturnMock = {
  limitData: {
    limit: {
      cd: {
        totalServiceInstances: 100,
        totalWorkload: 200
      }
    }
  },
  usageData: {
    usage: {
      cd: {
        activeServiceInstances: {
          count: 20,
          displayName: 'Last 30 Days'
        },
        activeServices: {
          count: 45,
          displayName: 'Last 30 Days'
        }
      }
    }
  }
}

describe('CDUsageInfo', () => {
  test('CDUsageInfo', () => {
    const { container } = render(
      <TestWrapper>
        <CDUsageInfo />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })
})
