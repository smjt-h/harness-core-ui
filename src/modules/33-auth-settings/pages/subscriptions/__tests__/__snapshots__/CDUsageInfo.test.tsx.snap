// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDUsageInfo CDUsageInfo 1`] = `
<div>
  <div
    class="Layout--horizontal Layout--layout-spacing-large StyledProps--main"
  >
    <div
      class="bp3-card bp3-elevation-0 Card--card innerCard"
    >
      <div
        class="Layout--vertical StyledProps--main"
      >
        <div
          class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-alignItems-baseline StyledProps--flex-justifyContent-space-between"
        >
          <div
            class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--flex StyledProps--flex-alignItems-baseline"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-grey700"
            >
              common.subscriptions.usage.srvcInst
            </p>
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="info"
                  tabindex="0"
                >
                  <svg
                    height="15"
                    viewBox="0 0 12 12"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M0 6a6 6 0 1012 0A6 6 0 000 6zm.75 0a5.25 5.25 0 1010.5 0A5.25 5.25 0 00.75 6zm4.206 3.104v.646h2.682v-.646l-.836-.144V4.203H4.956v.65l.836.144V8.96l-.836.144zm1.846-7.352v1.03h-1.01v-1.03h1.01z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </div>
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-grey500"
          >
            Last 30 Days
          </p>
        </div>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-size-large StyledProps--font-weight-bold StyledProps--color StyledProps--color-black"
        >
          20
        </p>
        <div
          class="StyledProps--font StyledProps--main bar StyledProps--background-grey100 StyledProps--border StyledProps--margin-bottom-small"
          style="width: 100%; height: 10px;"
        >
          <div
            class="StyledProps--font StyledProps--main StyledProps--background-primary6 StyledProps--border"
            style="width: 20%; height: 100%;"
          />
        </div>
        <div
          class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-size-xsmall"
          >
            100
             
            common.subscribed
          </p>
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-size-xsmall StyledProps--color"
          >
            20
            % 
            common.subscribed
          </p>
        </div>
      </div>
    </div>
    <div
      class="bp3-card bp3-elevation-0 Card--card innerCard"
    >
      <div
        class="Layout--vertical StyledProps--main"
      >
        <div
          class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-alignItems-baseline StyledProps--flex-justifyContent-space-between"
        >
          <div
            class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--flex StyledProps--flex-alignItems-baseline"
          >
            <p
              class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-grey700"
            >
              common.subscriptions.usage.services
            </p>
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="info"
                  tabindex="0"
                >
                  <svg
                    height="15"
                    viewBox="0 0 12 12"
                    width="15"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M0 6a6 6 0 1012 0A6 6 0 000 6zm.75 0a5.25 5.25 0 1010.5 0A5.25 5.25 0 00.75 6zm4.206 3.104v.646h2.682v-.646l-.836-.144V4.203H4.956v.65l.836.144V8.96l-.836.144zm1.846-7.352v1.03h-1.01v-1.03h1.01z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </div>
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-grey500"
          >
            Last 30 Days
          </p>
        </div>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-size-large StyledProps--font-weight-bold StyledProps--color StyledProps--color-black"
        >
          45
        </p>
        <div
          class="StyledProps--font StyledProps--main bar StyledProps--background-grey100 StyledProps--border StyledProps--margin-bottom-small"
          style="width: 100%; height: 10px;"
        >
          <div
            class="StyledProps--font StyledProps--main StyledProps--background-primary6 StyledProps--border"
            style="width: 22.5%; height: 100%;"
          />
        </div>
        <div
          class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-size-xsmall"
          >
            200
             
            total
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
